### PR TITLE
Rename GTasks Task list

### LINKED ISSUE #163: Google Tasks: Rename a Task List
## Problem

It should be possible to rename a task list for a Google backend.

## Proposal

Since there is no high level API in EDS for renaming tasks lists in the backend, we need to find a lower level API way to do this. See how its done for a regular WebDAV backend in PR #153

### PR DESCRIPTION
**Pull request by marbetschar**
This PR fixes #163

### CODE REVIEW

**Review by danrabbit — COMMENTED** [review]
This works as expected for renaming GTasks lists. I am encountering this error though:

```
(io.elementary.tasks:5319): GLib-GObject-WARNING **: 11:34:30.995: invalid cast from 'EGDataOAuth2Authorizer' to 'GDataOAuth2Authorizer'
```

**Comment by marbetschar** [comment]
@danrabbit that's probably due to [the following line in the code](https://github.com/elementary/tasks/blob/bc5222955815cfa1039e149e8abb1e0f26458bc1/src/TaskModel.vala#L185):

```
var oauth2_authorizer = (GData.OAuth2Authorizer) new E.GDataOAuth2Authorizer (collection_source, typeof (GData.TasksService));
```

tbh I don't know if that's the way how the code is intended to be used. As there is no unified, high level API for task list managing in libedataserver I'm doing a lot of guess work here. But I don't see why libedataserver should provide a `E.GDataOAuth2Authorizer` other than allow for authorization to use a libgdata backend - therefore casting it to a `GData.OAuth2Authorizer` seemed the way to go.

**Comment by marbetschar** [comment]
maybe @tintou knows, if this the correct way to glue `libgdata` and `libedataserver-1.2` together?

The below code produces a warning (`invalid cast from 'EGDataOAuth2Authorizer' to 'GDataOAuth2Authorizer'`), but other than that it works:

```
var oauth2_authorizer = (GData.OAuth2Authorizer) new E.GDataOAuth2Authorizer (collection_source, typeof (GData.TasksService));
```

**Comment by tintou** [comment]
@marbetschar I've now opened https://gitlab.gnome.org/GNOME/evolution-data-server/-/merge_requests/57
Just FYI, after this PR, the .vapi contains this:
```vala
public class GDataOAuth2Authorizer : GLib.Object, GData.Authorizer {
```
So you might want to cast to `GData.Authorizer` instead.

File: src/TaskModel.vala [hunk 1]
```diff
@@ -179,6 +179,32 @@ public class Tasks.TaskModel : Object {
...
+            unowned GLib.List<GData.Entry> gtasks_all_tasklist_entries = gtasks_service.query_all_tasklists (null, null, null).get_entries ();
+            foreach (var gtask_tasklist_entry in gtasks_all_tasklist_entries) {
+                if ("gtasks::%s".printf (gtask_tasklist_entry.id) == task_list_resource_extension.identity) {
+                    gtasks_tasklist = (GData.TasksTasklist) gtask_tasklist_entry;
+                    break;
+                }
+            }
+
+            if (gtasks_tasklist == null) {
+                throw new Tasks.TaskModelError.BACKEND_ERROR ("Task list '%s' is no longer available in Google backend.".printf (task_list_resource_extension.identity));
```
Comment (tintou) [inline src/TaskModel.vala:200]:
```suggestion
                throw new Tasks.TaskModelError.BACKEND_ERROR ("Task list '%s' is no longer available in Google backend.", task_list_resource_extension.identity);
```
Errors are printf by default

File: src/TaskModel.vala [hunk 2]
```diff
@@ -179,6 +179,32 @@ public class Tasks.TaskModel : Object {
...
+            debug ("GTasks Rename '%s'", task_list.get_uid ());
+
+            var authorizer = (GData.Authorizer) new E.GDataOAuth2Authorizer (collection_source, typeof (GData.TasksService));
+            var gtasks_service = new GData.TasksService (authorizer);
+            var task_list_resource_extension = (E.SourceResource) task_list.get_extension (E.SOURCE_EXTENSION_RESOURCE);
+
+            GData.TasksTasklist? gtasks_tasklist = null;
+
+            unowned GLib.List<GData.Entry> gtasks_all_tasklist_entries = gtasks_service.query_all_tasklists (null, null, null).get_entries ();
+            foreach (var gtask_tasklist_entry in gtasks_all_tasklist_entries) {
```
Comment (tintou) [inline src/TaskModel.vala:192]:
```suggestion
            foreach (unowned var gtask_tasklist_entry in gtasks_all_tasklist_entries) {
```

**Review by danrabbit — APPROVED** [review]
Works as expected. No more Terminal errors. Nice job :)

**Pull request merged**